File: frontend/types/career-guidance.ts
/**
 * Mirrors backend/app/ai/schemas/career_guidance.py exactly -- field-for-
 * field, same nullability. This is the Phase 6 Career Orchestrator +
 * Career Advisor response: every canonical status/skill_id/score/band
 * field is attached server-side from the same deterministic engines every
 * other student page already trusts (skill_gap_service, match_service,
 * learning_recommendation_service) -- nothing here is recomputed
 * client-side, and nothing here is unrestricted LLM prose. `headline`,
 * `reason`, and `suggested_action` text are server-authored or
 * server-grounded, never raw model output rendered without grounding.
 */

import type { OpportunityMatch, StudentOpportunityDetail } from "@/types/student-opportunity";
import type { YouTubeVideoCandidate } from "@/types/youtube-learning";

export type AnalysisMode = "JOB_ROLE" | "PERSONAL";

export type AssessmentActionType =
  | "ADD_SKILL_THEN_ASSESS"
  | "TAKE_ASSESSMENT"
  | "ALREADY_VERIFIED"
  | "NO_ASSESSMENT_AVAILABLE";

export type CareerActionCode =
  | "LEARN_SKILL"
  | "TAKE_ASSESSMENT"
  | "START_COURSE"
  | "APPLY_OPPORTUNITY"
  | "BUILD_PROJECT"
  | "REASSESS_SKILL";

/** Mirrors `CourseCandidate` (app.ai.schemas.course_recommendation) --
 * every field here is source-owned (the internal learning catalog, or a
 * future configured external provider); never LLM-authored. */
export interface CourseCandidate {
  candidate_id: string;
  source_type: "INTERNAL" | "EXTERNAL";
  provider: string | null;
  title: string;
  url: string;
  description: string | null;
  skill_ids: string[];
  skill_names: string[];
  level: string | null;
  duration_text: string | null;
  price_text: string | null;
  rating: number | null;
  certificate_available: boolean | null;
  external_course_id: string | null;
  metadata_source: string;
}

/** Mirrors `CareerSummary`. `headline` is the only AI-authored field on
 * this whole response and is null whenever Career Advisor synthesis did
 * not run -- never fabricated as a substitute for a canonical number. */
export interface CareerSummary {
  mode: AnalysisMode;
  target_role: string | null;
  readiness_score: number | null;
  headline: string | null;
}

/** Mirrors `PrioritySkill`. `reason`/`suggested_action` come from the
 * Skill Gap Agent's own grounded analysis (Phase 3) -- never re-authored
 * here. `highlighted_by_advisor` is the only field the Career Advisor
 * influences, and it only ever reorders/flags, never invents. */
export interface PrioritySkill {
  skill_id: string;
  skill_name: string;
  canonical_status: string | null;
  canonical_priority: string | null;
  canonical_importance: string | null;
  reason: string;
  suggested_action: string;
  highlighted_by_advisor: boolean;
}

export interface LearningRecommendationItem {
  course: CourseCandidate;
  for_skill_id: string | null;
  for_skill_name: string | null;
  reason: string;
  highlighted_by_advisor: boolean;
}

/** Mirrors `YouTubeRecommendationItem` -- additive sibling of
 * `LearningRecommendationItem`, never merged into it (a video is not a
 * CourseCandidate). Never touched by the Career Advisor -- see the
 * backend class's own docstring -- so there is no
 * `highlighted_by_advisor` field here. */
export interface YouTubeRecommendationItem {
  video: YouTubeVideoCandidate;
  for_skill_id: string | null;
  for_skill_name: string | null;
  reason: string;
}

export interface OpportunityRecommendationItem {
  opportunity: StudentOpportunityDetail;
  match: OpportunityMatch;
  reason: string;
  highlighted_by_advisor: boolean;
}

export interface AssessmentRecommendationItem {
  skill_id: string;
  skill_name: string;
  action: AssessmentActionType;
  assessment_available: boolean;
  note: string;
  highlighted_by_advisor: boolean;
}

/** `entity_id`/`entity_name` are always copied from the referenced
 * specialist's own grounded object server-side -- never from LLM text. */
export interface NextAction {
  order: number;
  type: CareerActionCode;
  entity_type: "SKILL" | "COURSE" | "OPPORTUNITY" | "ASSESSMENT";
  entity_id: string;
  entity_name: string;
}

export interface AgentStatus {
  available: boolean;
  fallback_used: boolean;
  detail: string | null;
}

export interface CareerGuidanceAgentsUsed {
  skill_gap_agent: AgentStatus;
  course_agent: AgentStatus;
  opportunity_agent: AgentStatus;
  career_advisor: AgentStatus;
}

export interface CareerGuidanceMeta {
  provider: "groq";
  model: string;
  ai_available: boolean;
  agents_used: CareerGuidanceAgentsUsed;
}

export interface CareerGuidanceResponse {
  career_summary: CareerSummary;
  priority_skills: PrioritySkill[];
  learning_recommendations: LearningRecommendationItem[];
  youtube_videos: YouTubeRecommendationItem[];
  opportunity_recommendations: OpportunityRecommendationItem[];
  assessment_recommendations: AssessmentRecommendationItem[];
  next_actions: NextAction[];
  meta: CareerGuidanceMeta;
  disclaimer: string;
}

File: backend/app/ai/schemas/career_guidance.py
"""Phase 6: Career Orchestrator + Career Advisor schemas.

Three groups, same separation as every earlier phase:

1. CareerAdvisorAgentInput -- minimal, ref-annotated input built ONLY
   from the three specialists' ALREADY-GROUNDED outputs (never raw
   student data, never a second StudentCareerContext fetch). Refs here
   are a FRESH namespace the Career Advisor owns (G#/C#/OP#/A#), built
   from the specialists' final response objects (skill_id /
   candidate_id / opportunity.id) -- not the same ref tokens those
   agents used internally with Groq, which do not survive into their
   final response objects.

2. CareerAdvisorLLMOutput -- exactly what Groq is asked to produce.
   Flat only: a headline string, three ref-list fields, and one
   "<ref>: ACTION_CODE" note list. No canonical field (score, band,
   status, priority, verification, price, url, ...) exists anywhere in
   this class.

3. The rich CareerGuidanceResponse -- what
   POST /api/v1/ai/career-guidance returns. Every canonical_*/
   skill_id/course/opportunity/action field is attached server-side by
   app.ai.guardrails.career_advisor + app.ai.orchestrator, reusing the
   specialists' own already-grounded objects verbatim -- never
   reconstructed from LLM text.
"""

from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.ai.schemas.base import AIResponseMeta
from app.ai.schemas.course_recommendation import CourseCandidate
from app.ai.schemas.skill_gap import AssessmentActionType
from app.ai.schemas.youtube_learning import YouTubeVideoCandidate
from app.schemas.skill_gap import AnalysisMode
from app.schemas.student_opportunity import OpportunityMatchResponse, StudentOpportunityDetail

_MAX_LIST_ITEMS = 10
_MAX_TEXT = 400

# ============================================================
# 1. Advisor input -- minimal, ref-annotated, built from specialist output
# ============================================================


class AdvisorInputSkill(BaseModel):
    ref: str
    skill_name: str
    canonical_status: str | None
    canonical_priority: str | None
    canonical_importance: str | None


class AdvisorInputCourse(BaseModel):
    ref: str
    title: str
    for_skill_ref: str | None


class AdvisorInputOpportunity(BaseModel):
    ref: str
    title: str
    opportunity_type: str
    match_band: str
    match_score: int


class AdvisorInputAssessmentAction(BaseModel):
    ref: str
    skill_name: str
    action: str


class CareerAdvisorAgentInput(BaseModel):
    """Never contains student_id, email, phone, or portfolio text --
    only refs and the small canonical fields needed to reason about
    them. Built by app.ai.orchestrator from the three specialists'
    OWN grounded outputs, never from a fresh context/DB read."""

    target_role: str | None
    readiness_score: int | None
    mode: AnalysisMode
    skills: list[AdvisorInputSkill] = Field(default_factory=list)
    courses: list[AdvisorInputCourse] = Field(default_factory=list)
    opportunities: list[AdvisorInputOpportunity] = Field(default_factory=list)
    assessment_actions: list[AdvisorInputAssessmentAction] = Field(default_factory=list)


# ============================================================
# 2. Raw LLM output -- flat, ref + action-code only
# ============================================================


class CareerAdvisorLLMOutput(BaseModel):
    """What Groq must produce. No skill_id, no score, no band, no
    status, no priority, no verification, no URL -- only refs, a short
    headline, and "<ref>: ACTION_CODE" notes validated server-side."""

    model_config = ConfigDict(extra="forbid")

    headline: str = Field(min_length=1, max_length=_MAX_TEXT)
    focus_skill_refs: list[str] = Field(
        default_factory=list,
        max_length=_MAX_LIST_ITEMS,
        description="Bare G# refs, most important first.",
    )
    course_refs: list[str] = Field(
        default_factory=list, max_length=_MAX_LIST_ITEMS, description="Bare C# refs to highlight."
    )
    opportunity_refs: list[str] = Field(
        default_factory=list, max_length=_MAX_LIST_ITEMS, description="Bare OP# refs to highlight."
    )
    action_notes: list[str] = Field(
        default_factory=list,
        max_length=_MAX_LIST_ITEMS,
        description=(
            'One string per suggested action, formatted exactly as "<ref>: ACTION_CODE" -- ref is '
            "any supplied G#/C#/OP#/A# token; ACTION_CODE is one of LEARN_SKILL, TAKE_ASSESSMENT, "
            "START_COURSE, APPLY_OPPORTUNITY, BUILD_PROJECT, REASSESS_SKILL."
        ),
    )
    learning_sequence: list[str] = Field(
        default_factory=list,
        max_length=_MAX_LIST_ITEMS,
        description="Bare G# refs in suggested learning order.",
    )


# ============================================================
# 3. Grounded, rich response -- canonical fields server-attached
# ============================================================


class ActionCode(str, Enum):
    LEARN_SKILL = "LEARN_SKILL"
    TAKE_ASSESSMENT = "TAKE_ASSESSMENT"
    START_COURSE = "START_COURSE"
    APPLY_OPPORTUNITY = "APPLY_OPPORTUNITY"
    BUILD_PROJECT = "BUILD_PROJECT"
    REASSESS_SKILL = "REASSESS_SKILL"


class CareerSummary(BaseModel):
    """Every field here comes from skill_gap_service's own canonical
    output (via the Skill Gap Agent's own `canonical` section) --
    `headline` is the only AI-authored field, and is None whenever the
    Career Advisor synthesis itself did not run."""

    mode: AnalysisMode
    target_role: str | None = None
    readiness_score: int | None = None
    headline: str | None = None


class PrioritySkill(BaseModel):
    skill_id: str
    skill_name: str
    canonical_status: str | None
    canonical_priority: str | None
    canonical_importance: str | None
    reason: str
    suggested_action: str
    highlighted_by_advisor: bool = False


class LearningRecommendationItem(BaseModel):
    course: CourseCandidate
    for_skill_id: str | None
    for_skill_name: str | None
    reason: str
    highlighted_by_advisor: bool = False


class YouTubeRecommendationItem(BaseModel):
    """Additive sibling of LearningRecommendationItem -- see
    app.ai.schemas.youtube_learning.YouTubeVideoCandidate's own
    docstring on why a video stays its own type rather than a
    CourseCandidate. Never touched by the Career Advisor: video is not
    part of its G#/C#/OP#/A# ref namespace, so `highlighted_by_advisor`
    does not apply here -- see app.ai.orchestrator."""

    video: YouTubeVideoCandidate
    for_skill_id: str | None
    for_skill_name: str | None
    reason: str


class OpportunityRecommendationItem(BaseModel):
    opportunity: StudentOpportunityDetail
    match: OpportunityMatchResponse
    reason: str
    highlighted_by_advisor: bool = False


class AssessmentRecommendationItem(BaseModel):
    skill_id: str
    skill_name: str
    action: AssessmentActionType
    assessment_available: bool
    note: str
    highlighted_by_advisor: bool = False


class NextAction(BaseModel):
    """One concrete, ranked next step. `entity_id`/`entity_name` are
    always copied from a specialist's own grounded object -- never from
    LLM text. `order` is server-assigned (either the Career Advisor's
    grounded sequence, or a deterministic priority-based fallback)."""

    order: int
    type: ActionCode
    entity_type: Literal["SKILL", "COURSE", "OPPORTUNITY", "ASSESSMENT"]
    entity_id: str
    entity_name: str


class AgentStatus(BaseModel):
    available: bool
    fallback_used: bool
    detail: str | None = None


class CareerGuidanceAgentsUsed(BaseModel):
    skill_gap_agent: AgentStatus
    course_agent: AgentStatus
    opportunity_agent: AgentStatus
    career_advisor: AgentStatus


class CareerGuidanceMeta(AIResponseMeta):
    ai_available: bool = False
    agents_used: CareerGuidanceAgentsUsed


class CareerGuidanceResponse(BaseModel):
    career_summary: CareerSummary
    priority_skills: list[PrioritySkill] = Field(default_factory=list)
    learning_recommendations: list[LearningRecommendationItem] = Field(default_factory=list)
    youtube_videos: list[YouTubeRecommendationItem] = Field(default_factory=list)
    opportunity_recommendations: list[OpportunityRecommendationItem] = Field(default_factory=list)
    assessment_recommendations: list[AssessmentRecommendationItem] = Field(default_factory=list)
    next_actions: list[NextAction] = Field(default_factory=list)
    meta: CareerGuidanceMeta
    disclaimer: str = (
        "This career plan combines your canonical skill, course, and opportunity data. "
        "AI synthesis is advisory only -- it never changes your readiness, verification, "
        "match scores, or any other canonical fact."
    )
